README.md
# Dialogue Generator for Musio interview

After installing the requirements by `pip install -r requirements.txt` command, edit the prompt text in `dialogue_prompt.txt` and paths, if needed.


1. `python3 generate_dialogues.py` will generate the dialogues according to the list of topics in `topics` folder (the topics can be added to different files). The dialogues are saved in the `dialogues` folder under the filename of their respective topics.

2. `python3 generate_summaries.py` generates summaries for each dialogue saved in `dialogues` folder.

3. `python3 QandA/dialogue_to_csv.py` collects every Q and A pairs between Musio and the student and save it in a csv (`delimiter = ';'`) file. Sample data in csv format:
  <p align="center">
    <img width="800" align = 'center' alt="Screen Shot 2022-08-08 at 8 08 40 PM" src="https://user-images.githubusercontent.com/108511037/183404723-b27f68ab-fb9b-46cd-b993-ba6c25a23e06.png">
  </p>

4. `python3 QandA/score_answers_gpt3.py` will score the answers from the students. The script reads Q and A pair from the csv file and scores the answer using OpenAI's api and writes it to another csv file.


dialogues/news
Musio: So, user, what do you think about the current state of the news?

User: Well, Musio, I think that the current state of the news is pretty bad. I mean, there's so much violence and hatred being reported on, it's hard to keep up with all of it. And it's not just in America, either - it seems like everywhere in the world, there's some kind of conflict going on.

Musio: That's a really interesting perspective. Do you think that the way the news is reported on might be part of the problem?

User: Absolutely. I think that a lot of times, the news focuses on all of the negative aspects of what's going on, instead of any of the good. And I think that can make people feel like there's no hope - like things will never get better.

Musio: That makes a lot of sense. So what do you think could be done to improve the current state of the news?

User: Well, I think that one thing that could be done is to focus more on stories that show how people are helping each other, or working together to make a difference. I think those kinds of stories would give people more hope, and make them feel like they can make a difference too.

Musio: That's a great idea. I'm sure a lot of people would love to see more positive stories in the news. Thank you for your time, user.

generate_summaries.py
import openai
from random import seed,sample
import os
import re


with open('openaiapikey.txt', 'r') as infile:
    open_ai_api_key = infile.read()
openai.api_key = open_ai_api_key


seed()
dialoguedir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/dialogues/'
outdir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/summaries/'
promptdir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/summary_prompt.txt'
files = os.listdir(dialoguedir)
# files = [i for i in files if 'stack' in i]  # filter list: dialog, medical, reddit, stack, news
# files = sample(files, 16)
print(files)

# davinci-instruct
# temp = 0.9
# top_p = 0.95
# freq pen = 0.5
# pres pen = 0.5


def load_prompt(filename, payload):
    with open(filename, 'r', encoding='utf-8') as infile:
        body = infile.read()
        body = body.replace('<<DIALOGUE>>', payload)
        return body


def completion(prompt, engine='davinci-instruct-beta', temp=0.8, top_p=0.95, tokens=600, freq_pen=0.5, pres_pen=0.5, stop=None):
    try:
        response = openai.Completion.create(
            engine=engine,
            prompt=prompt,
            temperature=temp,
            max_tokens=tokens,
            top_p=top_p,
            frequency_penalty=freq_pen,
            presence_penalty=pres_pen,
            stop=stop)
        text = response['choices'][0]['text'].strip()
        return text
    except Exception as oops:
        print('ERROR in completion function:', oops)

for f in files:
    try:
        with open(dialoguedir + f, 'r', encoding='utf-8') as infile:
            context = infile.read()
        prompt = load_prompt(promptdir, context)
        print('\n---------------------\n', prompt)
        summary = completion(prompt)
        print('\n', summary)
        with open(outdir + f, 'w', encoding='utf-8') as outfile:
            outfile.write(summary)
    except Exception as oops:
        print('ERROR in main loop:', f, oops)

QandA/dialogue_to_csv.py
import csv
import os

csv_filename = 'dialogue_data.csv'
dialoguedir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/dialogues/'

dialogues = os.listdir(dialoguedir)

def get_data(filename):
    with open(dialoguedir + filename, 'r') as file:
        data = file.read().strip()
        data = data.replace("\n", ' ')
        data = data.split("Musio:")
        clean_data = []
        for sent in data:
            if sent:
                if "user:" in sent:
                    sents = sent.split("user:")
                    for clean_sent in sents:
                        if clean_sent:
                            clean_data.append(clean_sent.strip())
                    continue
                elif "User:" in sent:
                    sents = sent.split("User:")
                    for clean_sent in sents:
                        if clean_sent:
                            clean_data.append(clean_sent.strip())
                    continue
                else:
                    clean_data.append(sent.strip())

    return clean_data

with open(csv_filename, 'a+') as file:
    writer = csv.writer(file, delimiter=';')
    dialogue_num = 1
    for dialogue in dialogues:
        for_csv = get_data(dialogue)
        ind = 0

        while ind + 1 < len(for_csv):
            writer.writerow([dialogue_num] + for_csv[ind: ind+2])
            ind += 2

        dialogue_num += 1



QandA/score_answers_gpt3.py
import openai
from random import seed
import pandas
import numpy as np


with open('../openaiapikey.txt', 'r') as infile:
    open_ai_api_key = infile.read()
openai.api_key = open_ai_api_key


seed(42)
csv_file_dir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/QandA/'
prompt_dir = "/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/QandA/"
csv_filename = 'dialogue_data.csv'
out_filename = 'dialogue_data_scored.csv'
prompt_name = 'score_answer_prompt.txt'



def load_prompt(filename, payload1, payload2):
    with open(prompt_dir + filename, 'r', encoding='utf-8') as infile:
        body = infile.read()
        body = body.replace('<<QUESTION>>', payload1)
        body = body.replace('<<ANSWER>>', payload2)
        return body


def completion(prompt, engine='text-davinci-002', temp=0.7, top_p=1, tokens=10, freq_pen=0.5, pres_pen=0.5,  stop=None):
    try:
        response = openai.Completion.create(
            engine=engine,
            prompt=prompt,
            temperature=temp,
            max_tokens=tokens,
            top_p=top_p,
            frequency_penalty=freq_pen,
            presence_penalty=pres_pen,
            stop=stop)
        text = response['choices'][0]['text'].strip()
        if text.isdigit():
            return text
        elif "Score:" in text:
            return text[-1]
        else:
            print(f"Response from GPT-3: {text}")
            raise ValueError("The returned score is not an integer. Please check the Prompt.")
    except Exception as oops:
        print('ERROR in completion function:', oops)



df = pandas.read_csv(csv_file_dir + csv_filename, delimiter=';')

num_data = len(df)
for row_num in range(num_data):

    if np.isnan(df.loc[row_num, " Score"]):
        question = df.loc[row_num, " Question"]
        answer = df.loc[row_num, " Answer"]

        prompt_body = load_prompt(prompt_name, question, answer)
        score = completion(prompt_body)

        df.loc[row_num, " Score"] = score
        print(f"{row_num}/{num_data} Completed")
    # else:
    #     print(f"{row_num} - Already scored!")


df.to_csv(csv_file_dir + out_filename, sep = ';', index=False)
   


generate_dialogues.py
import openai
from random import seed, sample
import os
import re


with open('openaiapikey.txt', 'r') as infile:
    open_ai_api_key = infile.read()
openai.api_key = open_ai_api_key


seed()
topicsdir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/topics/'
outdir = '/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/dialogues/'
files = os.listdir(topicsdir)
# files = [i for i in files if 'news' in i]    # filter list: dialog, medical, reddit, stack, news
prompt_name = 'dialogue_prompt.txt'
# files = sample(files, 1)
print(files)

# davinci-instruct
# temp = 0.9
# top_p = 0.95
# freq pen = 0.5
# pres pen = 0.5


def load_prompt(filename, payload):
    with open('/Users/webby/Desktop/aka-files/working-dir/NLCA_Question_Generator/%s' % filename, 'r', encoding='utf-8') as infile:
        body = infile.read()
        body = body.replace('<<TOPIC>>', payload)
        return body


def completion(prompt, engine='text-davinci-002', temp=0.7, top_p=1, tokens=600, freq_pen=0.5, pres_pen=0.5,  stop=None):
    try:
        response = openai.Completion.create(
            engine=engine,
            prompt=prompt,
            temperature=temp,
            max_tokens=tokens,
            top_p=top_p,
            frequency_penalty=freq_pen,
            presence_penalty=pres_pen,
            stop=stop)
        text = response['choices'][0]['text'].strip().splitlines()
        print("text:",text)
        questions = ''
        for t in text:
            questions += re.sub('^\-', '', t).strip() + '\n'
        questions = questions.strip()
        return questions
    except Exception as oops:
        print('ERROR in completion function:', oops)

for f in files:
    try:
        with open(topicsdir + f, 'r', encoding='utf-8') as infile:
            topics = infile.read()
            topics = topics.split('\n')
        for topic in topics:
            topic = topic.strip()
            if not topic or '#' in topic: # skip over topics that are already used
                continue
            try:
                prompt = load_prompt(prompt_name, topic)
                print('\n---------------------\n', prompt)
                questions = completion(prompt)
                print('\n---------------------\n', questions)
                with open(outdir + topic, 'w', encoding='utf-8') as outfile:
                    outfile.write('Musio: ')
                    outfile.write(questions)
            except Exception as oops1:
                print("Error while requesting GPT-3", f, oops1)
    except Exception as oops:
        print('ERROR in main loop:', f, oops)